### PR TITLE
BlockLinearOperator: Allow identical destination and source

### LINKED ISSUE #6437: Bug: Use intermediate storage for BlockLinearOperator<..>.vmult()
If source and destination vectors are the same, the following code in `block_linear_operator.h` garbles up the result:
```
359         for (unsigned int i = 0; i < m; ++i)                                     
360           {                                                                      
361             op.block(i, 0).vmult(v.block(i), u.block(0));                        
362             for (unsigned int j = 1; j < n; ++j)                                 
363               op.block(i, j).vmult_add(v.block(i), u.block(j));                  
364           } 
```
We really have to use some intermediate storage here.
<s>Should be not much of a deal - I plan to fix this issue between rc1 and rc2.</s>

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #6437. We just provide a `apply_with_intermediate_storage` similar to the one for `LinearOperator`.
This PR also fixes a bug in the initialization of the `vmult` method for `MatrixInterfaceWithVmultAdd` objects.

### CODE REVIEW

File: include/deal.II/lac/block_linear_operator.h [hunk 1]
```diff
@@ -351,6 +351,43 @@ namespace internal
 {
   namespace BlockLinearOperatorImplementation
```
Comment (tamiko) [inline include/deal.II/lac/block_linear_operator.h]: Can you put these two helper functions in an anonymous namespace please, i.e.
```
namespace 
  {
    // helper functions...
  }
```

We have `internal::LinearOperatorImplementation` and `internal::BlockLinearOperatorImplementation` for things that need to be overriden in other header files (reinit, payload). The corresponding `apply_with_intermediate_storage` etc. is in an anonymous namespace.

File: include/deal.II/lac/block_linear_operator.h [hunk 2]
```diff
@@ -384,96 +421,138 @@ namespace internal
...
+        if (PointerComparison::equal(&v, &u))
           {
-            op.block(i, 0).vmult(v.block(i), u.block(0));
-            for (unsigned int j = 1; j < n; ++j)
+            const auto first_op = [&op](Range &            v,
+                                        const Domain &     u,
+                                        const unsigned int i,
+                                        const unsigned int j) {
+              op.block(i, j).vmult(v.block(i), u.block(j));
+            };
```
Comment (tamiko) [inline include/deal.II/lac/block_linear_operator.h]: I am a huge fan of lambda captures. :-)

File: include/deal.II/lac/block_linear_operator.h [hunk 3]
```diff
@@ -347,137 +347,218 @@ class BlockLinearOperator
 
 
 
-namespace internal
+namespace
```
Comment (bangerth) [inline include/deal.II/lac/block_linear_operator.h]: No, bad idea -- we don't want to use anonymous namespaces in header files. Can you please undo this change?

File: include/deal.II/lac/block_linear_operator.h [hunk 4]
```diff
@@ -567,8 +648,7 @@ block_operator(const BlockMatrixType &block_matrix)
     return BlockType(block_matrix.block(i, j));
   };
 
-  internal::BlockLinearOperatorImplementation::
-    populate_linear_operator_functions(return_op);
+  populate_linear_operator_functions(return_op);
```
Comment (bangerth) [inline include/deal.II/lac/block_linear_operator.h]: @tamiko You'll then also have to reinstate these namespace qualifications.

**Comment by tamiko** [comment]
On Thu, Jun 14, 2018, at 00:54 CDT, Wolfgang Bangerth <notifications@github.com> wrote:

> bangerth commented on this pull request.
>
>
>
>> @@ -347,137 +347,218 @@ class BlockLinearOperator
>  
>  
>  
> -namespace internal
> +namespace
>
> No, bad idea -- we don't want to use anonymous namespaces in header
> files. Can you please undo this change?

Why not - we do this in `linear_operator.h` as well for entirely
internal helper functions.

**Comment by tamiko** [comment]
@bangerth 
>  No, bad idea -- we don't want to use anonymous namespaces in header files. Can you please undo this change?

Why not - we do this in `linear_operator.h` as well for entirely internal helper functions.

The rationale is that `linear_operator.h` and `block_linear_operator.h` already contain all function and class definitions. Meaning that in every compilation unit there will be a separate instance (with weak linkage) anyway.

**Comment by masterleinad** [comment]
I just like to be consistent. Either we allow these everywhere or fix all instances.

I already started avoiding unnamed namespaces in header files and will provide a PR. If we decide to discard it, I can live with it.

**Comment by tamiko** [comment]
@bangerth @masterleinad Well, we can make that a more global effort. Ultimately, we should compile a  list of rules what you actually want to see. For example, we have quite a number of
```
namespace internal
{
  namespace SymmetricTensorImplementation
  {
    namespace
    {
      // templated function
    }
  }
}
```
floating around in `*.h` files that look a bit silly. I have no problem promoting those to `internal::SymmetricTensorImplementation`. On the other hand we should definitely keep all anonymous namespaces in `*.templates.h` files - there they mainly suppress symbol visibility (which is a good thing).

**Comment by masterleinad** [comment]
See #6759.

**Comment by bangerth** [comment]
The issue with anonymous namespaces is that they are defined as given a unique name in every translation unit. That means that if you have
```
  namespace { int foo () {} }
```
in a header file, then this function will show up in one object file as the mangled form of something along the lines of `abc::foo()` and in another object file as `def::foo()`. They both have internal linkage, but because they have different names, the compiler can not eliminate duplicates. So you get code bloat at the library/executable level.

But the main reason to not use anonymous namespaces is because inevitably someone thinks that it's a good idea to put a variable in there:
```
  namespace {
    int i;
  }
```
This looks innocuous, but in reality, you now have as many variables as the number of times this header file is included elsewhere. (If you mark the variable as `const` and initialize it in one `.cc` file, then all of the source files that include the header will yield a linker error. Same if you make the variable `extern`.)

**Comment by masterleinad** [comment]
Let us not discuss this in a closed PR. I'll move the discussion over to #6759.
